### ISSUE TITLE
merging volumes result in duplication and error

### ISSUE THREAD

**Issue by debrutal** [containers/podman-compose#256]
**Error description**

Scenario: File 1 describes volumes, file 2 overwrites that volume to be compatible with selinux, since they share the mounted file. 
Executing it with ``` podman-compose -f file1.yml -f file2.yml up```
Take a look at the merge, which creates duplication of the entries

File 1 (Main compose file)
```version: '3.7'

services:
  hello:
    image: hello-world
    volumes:
      -
        type: volume
        source: sample-volume
        target: /var/lib/
        read_only: false
      -
        type: bind
        source: ./text.txt
        target: /etc/config/text

  hello2:
    depends_on:
      - hello
    image: hello-world
    volumes:
      -
        type: volume
        source: sample-volume 
        target: /var/lib/
        read_only: false
      -
        type: bind
        source: ./text.txt
        target: /etc/config/text
volumes:
  sample-volume:
```

File 2 (Additional file adding selinux specific suffixes.)
```
version: '3.7'

services:
  hello:
    volumes:
      -
        type: volume
        source: sample-volume
        target: /var/lib/:z
        read_only: false
      -
        type: bind
        source: ./text.txt
        target: /etc/config/text

  hello2:
    volumes:
      -
        type: volume
        source: sample-volume 
        target: /var/lib/:z
        read_only: false
      -
        type: bind
        source: ./text.txt
        target: /etc/config/text


```
Result
```
using podman version: podman version 2.1.1
 ** merged:
 {
  "_dirname": "/home/dennis/sample",
  "version": "3.7",
  "services": {
    "hello": {
      "image": "hello-world",
      "volumes": [
        {
          "type": "volume",
          "source": "sample-volume",
          "target": "/var/lib/",
          "read_only": false
        },
        {
          "type": "bind",
          "source": "./text.txt",
          "target": "/etc/config/text"
        },
        {
          "type": "volume",
          "source": "sample-volume",
          "target": "/var/lib/:Z",
          "read_only": false
        },
        {
          "type": "bind",
          "source": "./text.txt",
          "target": "/etc/config/text"
        }
      ]
    },
    "hello2": {
      "depends_on": [
        "hello"
      ],
      "image": "hello-world",
      "volumes": [
        {
          "type": "volume",
          "source": "sample-volume",
          "target": "/var/lib/",
          "read_only": false
        },
        {
          "type": "bind",
          "source": "./text.txt",
          "target": "/etc/config/text"
        },
        {
          "type": "volume",
          "source": "sample-volume",
          "target": "/var/lib/:Z",
          "read_only": false
        },
        {
          "type": "bind",
          "source": "./text.txt",
          "target": "/etc/config/text"
        }
      ]
    }
  },
  "volumes": {
    "sample-volume": null
  }
}
podman pod create --name=sample --share net
ca990e2abbd0f68c680dabc25efd898555f270ea5a94f8d86e1acc3df28bb9a2
0
podman volume inspect sample_sample-volume || podman volume create sample_sample-volume
Error: error inspecting volume sample_sample-volume: no volume with name "sample_sample-volume" found: no such volume
podman volume inspect sample_sample-volume || podman volume create sample_sample-volume
podman create --name=sample_hello_1 --pod=sample --label io.podman.compose.config-hash=123 --label io.podman.compose.project=sample --label io.podman.compose.version=0.0.1 --label com.docker.compose.container-number=1 --label com.docker.compose.service=hello -v sample_sample-volume:/var/lib/:rw -v ./text.txt:/etc/config/text -v sample_sample-volume:/var/lib/:Z:rw -v ./text.txt:/etc/config/text --add-host hello:127.0.0.1 --add-host sample_hello_1:127.0.0.1 --add-host hello2:127.0.0.1 --add-host sample_hello2_1:127.0.0.1 hello-world
Trying to pull registry.fedoraproject.org/hello-world...
  manifest unknown: manifest unknown
Trying to pull registry.access.redhat.com/hello-world...
  name unknown: Repo not found
Trying to pull registry.centos.org/hello-world...
  manifest unknown: manifest unknown
Trying to pull docker.io/library/hello-world...
Getting image source signatures
Copying blob 0e03bdcc26d7 done  
Copying config bf756fb1ae done  
Writing manifest to image destination
Storing signatures
Error: sample_sample-volume:/var/lib/:Z:rw: incorrect volume format, should be [host-dir:]ctr-dir[:option]
125
podman volume inspect sample_sample-volume || podman volume create sample_sample-volume
podman volume inspect sample_sample-volume || podman volume create sample_sample-volume
podman create --name=sample_hello2_1 --pod=sample --label io.podman.compose.config-hash=123 --label io.podman.compose.project=sample --label io.podman.compose.version=0.0.1 --label com.docker.compose.container-number=1 --label com.docker.compose.service=hello2 -v sample_sample-volume:/var/lib/:rw -v ./text.txt:/etc/config/text -v sample_sample-volume:/var/lib/:Z:rw -v ./text.txt:/etc/config/text --add-host hello:127.0.0.1 --add-host sample_hello_1:127.0.0.1 --add-host hello2:127.0.0.1 --add-host sample_hello2_1:127.0.0.1 hello-world
Error: sample_sample-volume:/var/lib/:Z:rw: incorrect volume format, should be [host-dir:]ctr-dir[:option]
125
podman start -a sample_hello_1
Error: no container with name or ID sample_hello_1 found: no such container
125
podman start -a sample_hello2_1
Error: no container with name or ID sample_hello2_1 found: no such container
125
```

**Expected Behaviour**

should run without any errors. 

Background:
SELinux enabled

```
$ sestatus
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
SELinux root directory:         /etc/selinux
Loaded policy name:             targeted
Current mode:                   enforcing
Mode from config file:          enforcing
Policy MLS status:              enabled
Policy deny_unknown status:     allowed
Memory protection checking:     actual (secure)
Max kernel policy version:      33
```

**Comment by bhavin192** [comment]
There are multiple issues here, let me try to explain them one by one.


## 1. Support for `ACCESS_MODE` in long syntax of volumes

From the Docker compose documentation and the Compose spec, it seems that the long syntax (the one which is used in the above example) does not support the SELinux options i.e. `ACCESS_MODE`. So, I'm not sure if podman-compose should / can support it or not. That's why it ends up in something like `/var/lib/:Z:rw`.

-   [Compose file version 3 reference | Docker Documentation](https://docs.docker.com/compose/compose-file/compose-file-v3/#volumes)
-   [compose-spec/compose-spec/spec.md](https://github.com/compose-spec/compose-spec/blob/master/spec.md#volumes)

> The &#x2013;mount flag does not support z or Z options for modifying selinux labels.

[Use bind mounts | Docker Documentation](https://docs.docker.com/storage/bind-mounts/#choose-the--v-or---mount-flag)


## 2. Do volumes of type `volume` need the `z` label?

I never had to use the `:z` or `:Z` for volumes of type `volume`. Do you get any errors if you don't specify the `ACCESS_MODE` for these?


## 3. The volumes should get merged correctly

From the code of podman-compose's, I can see we are not merging the volumes the same way docker-compose does.

Depending of the destination value, the items from the list should get merged instead of simple concatenation.

> In the case of environment, labels, volumes, and devices, Compose “merges” entries together with locally-defined values taking precedence. For environment and labels, the environment variable or label name determines which value is used:

> Entries for volumes and devices are merged using the mount path in the container:

[Adding and overriding configuration](https://docs.docker.com/compose/extends/#adding-and-overriding-configuration)

This particular issue is about volumes but other directives mentioned above need to be handled as well.

I faced the 3rd issue yesterday, when I was trying to create a `docker-compose.override.yml`. Depending on what maintainers say about it, I might consider working on this.

**Comment by debrutal** [comment]
Thank you for differentiating each issue. 

My main focus was actually the 3rd one, since i would address this as a merging error. 


Background for issue 2:
There is a database configuration file, which is mounted in 2 containers. 
1st container is a Database-Migration container
2nd container is the application itself using the same configuration.
Since they share the same file, this crashes if SELinux is enabled. 

I didn't have this issue with docker, but i have to look into that again, to give more infos on the actual behaviour.

The actual idea behind this was:
I was trying to have an example on which i can use a few compose files. Something like 
- base-compose.yml
- with-selinux-enabled.yml
which can be easily overwrite / merge the mountings.
Also having a development team with heterogenous os configuration would require this to work somehow or have a workaround.

**Comment by muayyad-alsadi** [comment]
fixed as part of #333 
please test

**Comment by bhavin192** [comment]
@muayyad-alsadi it seems to work for the example mentioned in #333 but fails with following files:

docker-compose.yaml
```yaml
version: "3.3"
services:
  plausible_db:
    image: docker.io/postgres:12
    volumes:
      - db-data:/var/lib/postgresql/data
```

docker-compose.override.yaml
```yaml
version: "3.3"
services:
  plausible_db:
    restart: always
    volumes:
      - ${HOME}/data/db-data:/var/lib/postgresql/data:z
```

Results in:
```json
    "plausible_db": {
      "image": "docker.io/postgres:12",
      "volumes": [
        "db-data:/var/lib/postgresql/data",
        "/home/bhavin/data/db-data:/var/lib/postgresql/data:z"
      ],
      "restart": "always"
    },
```

The extra flags at the end need to be ignored while merging i.e. things like `ro`, `rw`, `ro,z` etc.
